### PR TITLE
Use shortest possible serialization for grid-area/grid-row/grid-column.

### PR DESCRIPTION
**Pull request by mattwoodrow**
#### fbf76efc64df0a981452a2387920cdbf90360cd8
<pre>
Use shortest possible serialization for grid-area/grid-row/grid-column.
<a href="https://bugs.webkit.org/show_bug.cgi?id=245026">https://bugs.webkit.org/show_bug.cgi?id=245026</a>

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/parsing/grid-area-valid-expected.txt:
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::isCustomIdentValue):
(WebCore::canOmitTrailingGridAreaValue):
(WebCore::StyleProperties::getGridRowColumnShorthandValue const):
(WebCore::StyleProperties::getGridAreaShorthandValue const):
* Source/WebCore/css/StyleProperties.h:

Canonical link: <a href="https://commits.webkit.org/254942@main">https://commits.webkit.org/254942@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f20728a66657c95e90d1372154be6c2514461146

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90783 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35360 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/21354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/100078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/158594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94791 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33859 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/83125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/96464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96438 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/77588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/83125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81724 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/21354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/83125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34947 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/21354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32750 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/21354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3446 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36524 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/77588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38449 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/21354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | |

### CODE REVIEW

File: Source/WebCore/css/StyleProperties.cpp [hunk 1]
```diff
@@ -997,9 +997,58 @@ String StyleProperties::getGridValue() const
...
+
 String StyleProperties::getGridShorthandValue(const StylePropertyShorthand& shorthand) const
 {
-    return getShorthandValue(shorthand, " / ");
+    // Look backwards through the longhands to see if there are any trailing
+    // values we can omit.
+    unsigned last = shorthand.length() - 1;
+    for (; last > 0; --last) {
+        auto property = shorthand.properties()[last];
+        auto value = getPropertyCSSValue(property);
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp]: I worry this is using `getPropertyCSSValue` too many times, which can be slow if the vector of properties is big.
I would prefer to call it only once per longhand at the beginning, and then use some logic like Firefox:

```rust
    pub fn can_omit(&self, other: &Self) -> bool {
        if self.is_ident_only() {
            self == other
        } else {
            other.is_auto()
        }
    }
```

```rust
        fn to_css<W>(&self, dest: &mut CssWriter<W>) -> fmt::Result where W: fmt::Write {
            self.grid_row_start.to_css(dest)?;
            let mut trailing_values = 3;
            if self.grid_column_start.can_omit(self.grid_column_end) {
                trailing_values -= 1;
                if self.grid_row_start.can_omit(self.grid_row_end) {
                    trailing_values -= 1;
                    if self.grid_row_start.can_omit(self.grid_column_start) {
                        trailing_values -= 1;
                    }
                }
            }
            let values = [&self.grid_column_start, &self.grid_row_end, &self.grid_column_end];
            for value in values.iter().take(trailing_values) {
                dest.write_str(" / ")?;
                value.to_css(dest)?;
            }
            Ok(())
        }
```

**Review by Loirooriol — CHANGES_REQUESTED** [review]

File: Source/WebCore/css/StyleProperties.cpp (hunk 1 above)
Comment (mattwoodrow) [reply Source/WebCore/css/StyleProperties.cpp]: Alright, I can do that! I can't think of a nice way to handle all 3 properties with a single function with that approach, but probably not a big deal to duplicate.

File: Source/WebCore/css/StyleProperties.cpp [hunk 2]
```diff
@@ -1002,6 +1002,72 @@ String StyleProperties::getGridShorthandValue(const StylePropertyShorthand& shor
...
+
+static bool canOmitTrailingGridAreaValue(CSSValue& value, CSSValue& trailing)
+{
+    if (isCustomIdentValue(value)) {
+        if (isCustomIdentValue(trailing) && value.cssText() == trailing.cssText())
+            return true;
+    } else if (isValueID(trailing, CSSValueAuto))
+        return true;
+
+    return false;
```
Comment (Loirooriol) [inline Source/WebCore/css/StyleProperties.cpp]: Nit: can be shortened to

```cpp
    if (isCustomIdentValue(value))
        return isCustomIdentValue(trailing) && value.cssText() == trailing.cssText();
    return isValueID(trailing, CSSValueAuto);
```

**Review by Loirooriol — APPROVED** [review]
Informal LGTM

**Review by nt1m — APPROVED** [review]
Thanks Oriol! LGTM as well!

**Pull request closed without merging**